did-doc-diff/src/index.ts

```diff
@@ -0,0 +1,217 @@
+    key: key.did(),
+    sig: '',
+  }
+  const signed_tick = sign(tick, key)
+  return signed_tick
+}
+
+function authorise_key(doc: { [index: string]: any }, didkey: string) {
+  const valid = Boolean(
+    (doc['adx/account_keys'] && doc['adx/account_keys'].includes(didkey)) ||
+      (doc['adx/recovery_keys'] && doc['adx/recovery_keys'].includes(didkey)),
+  )
+  console.log('authorise_key:', valid)
+  return true
+}
+
+export async function diffs_to_did_doc(diffs: { [index: string]: any }) {
+  if (typeof diffs !== 'object') {
+    return undefined
+  }
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined
+  }
+  let last = tids[0]
+  let doc = diffs[last]
+
+  // add the id feild
+  doc['id'] =
+    'did:aic:' +
+    (await pid(uint8arrays.fromString(JSON.stringify(diffs[last]))))
+
+  for (const tid of tids.slice(1)) {
+    const diff = diffs[tid]
+    if (!(await validate_sig(diff))) {
+      console.log('Warning ignoreing: bad sig')
+      continue // not valid ignore it
+    }
+    if (!authorise_key(doc, diff.key)) {
+      console.log('Warning ignoreing: bad key')
+      continue // not valid ignore it
+    }
+
+    const { prev, patches } = diff
+    if (prev !== last) {
+      console.log('Warning ignoreing: out of order')
+      continue
+    }
+    doc = patch(doc, patches)
+    last = tid
+  }
+  return doc
+}
+
+const S32_CHAR = '234567abcdefghijklmnopqrstuvwxyz'
+const B32_CHAR = 'abcdefghijklmnopqrstuvwxyz234567'
+
+function s32encode(data: Uint8Array): string {
+  // this is gross so many alocatons and funtion calls :(
+  const base32 = uint8arrays.toString(data, 'base32')
+  const sort_order_invariant_base32 = base32.split('').map((c) => {
+    return S32_CHAR[B32_CHAR.indexOf(c)]
+  })
+  return sort_order_invariant_base32.join('')
+}
+
+async function pid(data: Uint8Array) {
+  // return the pid120 a 24 char pid
+  const hash = s32encode((await sha256.digest(data)).digest)
+  let twos = 0
+  for (twos = 0; twos < 32; twos++) {
+    if (hash[twos] !== '2') {
+      break
+    }
+  }
+  const head = S32_CHAR[31 - twos]
+  const tail = hash.slice(twos, twos + 15) // 23 for pid120; 15 for pid80 // Are we ok with 80 bits?
+  return head + tail
+}
+
+function patch(doc: { [index: string]: any }, patches: Array<any>) {
+  for (const p of patches) {
+    const [op, path, value] = p
+    let node = doc
+    for (const segment of path.slice(0, -1)) {
+      if (op === 'put' && !(segment in node)) {
+        node[segment] = {}
+      }
+      node = node[segment]
+    }
+    if (op === 'put') {
+      node[path[path.length - 1]] = value
+    }
+    if (op === 'del') {
+      delete node[path[path.length - 1]]
+    }
+  }
+  return doc
+}
+
+export async function sign(
+  doc: {
+    sig: string
+    key: string
+    [index: string]: any
+  },
+  key: crypto.EcdsaKeypair,
+) {
+  if (doc.key !== key.did()) {
+    throw 'passed in secret key and did do not match'
```

Comment:
wrap this in an error:
```ts
throw new Error('passed in secret key and did do not match')
```

Javascript stupidly lets you throw whatever you want. But we should try to only throw errors